runtime/vm/jfr.cpp

```diff
@@ -1152,6 +1173,167 @@ jfrDump(J9VMThread *currentThread, BOOLEAN finalWrite)
+ * the ones needed to bootstrap JFR JCL initialization.
+ */
+jlong
+getKnownJFREventType(const J9UTF8 *className)
+{
+	jlong result = INVALID_TYPE_ID;
+	if (J9UTF8_LITERAL_EQUALS_UTF8(className, "boolean")) {
+		result = BOOLEAN_TYPE_ID;
+	} else if (J9UTF8_LITERAL_EQUALS_UTF8(className, "byte")) {
+		result = BYTE_TYPE_ID;
+	} else if (J9UTF8_LITERAL_EQUALS_UTF8(className, "char")) {
+		result = CHAR_TYPE_ID;
+	} else if (J9UTF8_LITERAL_EQUALS_UTF8(className, "short")) {
+		result = SHORT_TYPE_ID;
+	} else if (J9UTF8_LITERAL_EQUALS_UTF8(className, "int")) {
+		result = INT_TYPE_ID;
+	} else if (J9UTF8_LITERAL_EQUALS_UTF8(className, "float")) {
+		result = FLOAT_TYPE_ID;
+	} else if (J9UTF8_LITERAL_EQUALS_UTF8(className, "long")) {
+		result = LONG_TYPE_ID;
+	} else if (J9UTF8_LITERAL_EQUALS_UTF8(className, "double")) {
+		result = DOUBLE_TYPE_ID;
+	} else if (J9UTF8_LITERAL_EQUALS_UTF8(className, "jdk/types/StackTrace")) {
+		result = STACKTRACE_TYPE_ID;
+	}
+
+	return result;
+}
+
+jlong
+getTypeIDUTF8(J9VMThread *currentThread, const J9UTF8 *className)
+{
+	jlong result = INVALID_TYPE_ID;
+
+	Trc_VM_getTypeIDUTF8_Entry(currentThread, J9UTF8_LENGTH(className), J9UTF8_DATA(className));
+
+	J9Class *clazz = hashClassTableAt(currentThread->javaVM->systemClassLoader, J9UTF8_DATA(className), J9UTF8_LENGTH(className));
+
+	if (NULL != clazz) {
+		result = getTypeID(currentThread, clazz);
+	} else {
+		result = getKnownJFREventType(className);
+	}
+
+	Trc_VM_getTypeIDUTF8_Exit(currentThread, J9UTF8_LENGTH(className), J9UTF8_DATA(className), clazz, result);
+
+	return result;
+}
+
+jlong
+getTypeID(J9VMThread *currentThread, J9Class *clazz)
+{
+	J9JavaVM *vm = currentThread->javaVM;
+	J9HashTable *typeIDTable = clazz->classLoader->typeIDs;
+	jlong result = INVALID_TYPE_ID;
+	J9JFRTypeID entry = {0};
+	J9JFRTypeID *jfrTypeID = &entry;
+
+	Trc_VM_getTypeID_Entry(currentThread, clazz);
+
+	Assert_VM_mustHaveVMAccess(currentThread);
+
+	omrthread_monitor_enter(vm->jfrState.typeIDMonitor);
+
+	if (NULL == typeIDTable) {
+		PORT_ACCESS_FROM_JAVAVM(vm);
+
+		typeIDTable = hashTableNew(OMRPORT_FROM_J9PORT(privatePortLibrary),
+								J9_GET_CALLSITE(),
+								0,
+								sizeof(J9JFRTypeID),
+								sizeof(J9JFRTypeID *),
+								0,
+								J9MEM_CATEGORY_CLASSES,
+								jfrTypeIDHashFn,
+								jfrTypeIDHashEqualFn,
+								NULL,
+								NULL);
+
+		if (NULL == typeIDTable) {
+			setNativeOutOfMemoryError(currentThread, 0, 0);
+			goto done;
+		}
+		clazz->classLoader->typeIDs = typeIDTable;
+	}
+
+	jfrTypeID->clazz = clazz;
+	jfrTypeID = (J9JFRTypeID *)hashTableFind(typeIDTable, jfrTypeID);
+
+	if (NULL == jfrTypeID) {
+		jfrTypeID = &entry;
+		jfrTypeID->id = vm->jfrState.typeIDcount;
+
+		vm->jfrState.typeIDcount += 1;
+
+		jfrTypeID = (J9JFRTypeID *)hashTableAdd(typeIDTable, jfrTypeID);
+		if (NULL == jfrTypeID) {
+			setNativeOutOfMemoryError(currentThread, 0, 0);
+			goto done;
+		}
+	}
+	result = jfrTypeID->id;
```

Comment:
Should this not handle reserved classes (e.g. `boolean.class`)?

Reply:
I dont think I follow, please elaborate?

Reply:
There is a `J9Class` for `boolean.class` (and other "reserved" types); if this is called with a pointer one of those `J9Class` objects, this will not answer the predefined type ID, but something 500 or above.

Reply:
Only classes that subclass `jdk.jfr.event.Event` will be called with `getTypeId(Class ...)` and none of the reserved types will be found in the bootloader classtable.

Reply:
Perhaps a comment pointing out that expectation would be helpful.